FUSETOOLS2-1372 - adding a command to create and open an adm file

lhein opened the pull request:
Signed-off-by: Lars Heinemann <lhein.smx@gmail.com>

apupier commented on package.json:
```diff
@@ -54,6 +54,11 @@
 		},
+		"commands": {
+			"category": "AtlasMap",
+			"title": "Create AtlapMap file",
```
```suggestion
			"title": "Create AtlasMap file",
```

apupier commented on src/extension.ts:
```diff
@@ -56,6 +58,21 @@ export function deactivate(context: vscode.ExtensionContext) {
+async function createAndOpenADM() {
+	const selectedWorkspaceFolder: vscode.WorkspaceFolder | undefined = await vscode.window.showWorkspaceFolderPick( {placeHolder: 'Please select the workspace folder in which the new file will be created.'} );
+	if (selectedWorkspaceFolder) {
+		await vscode.window.showInputBox( {placeHolder: "Enter the name of the new AtlasMap file"})
+			.then( async (fileName: string) => {
+				var file: string = `${selectedWorkspaceFolder.uri.fsPath}/${fileName}`;
+				if (!file.toLowerCase().endsWith('.adm')) {
+					file += '.adm';
+				}
+				await vscode.workspace.fs.writeFile(vscode.Uri.file(file), Buffer.from(''));
+				await vscode.commands.executeCommand('vscode.open', vscode.Uri.file(file));
+			});
```
using the wait notation everywhere will bemore readable, somethign like:
```suggestion

		const fileName = await vscode.window.showInputBox( {placeHolder: "Enter the name of the new AtlasMap file"});
		var file: string = `${selectedWorkspaceFolder.uri.fsPath}/${fileName}`;
		if (!file.toLowerCase().endsWith('.adm')) {
			file += '.adm';
		}
		await vscode.workspace.fs.writeFile(vscode.Uri.file(file), Buffer.from(''));
		await vscode.commands.executeCommand('vscode.open', vscode.Uri.file(file));

```

apupier reviewed (changes requested):
- in case of already existing files, the file is overwritten
- need a test
- either in this PR or in another one but need to add the telemetry

lhein commented:
First iteration adding only a global command. In the next step it might be handy to have it also in the context menu of the file explorer in VS Code.

lhein commented:
implemented the suggested fixes and added a file exists check as well as the telemetry event and added the change to the changeLog....test is still missing

apupier commented on src/extension.ts:
```diff
@@ -50,12 +52,44 @@ function sendOpenEvent() {
 }
 
+async function createAndOpenADM() {
+	sendCreateEvent();
```
I think it is better to send the create event at the end of the function for these reasons:
- in case of failure, the main functionality is not completely broken
- in case it is taking a non negligeable time, it is not slowing down the main functionality
- it allows to send the element only when the command is really called successfully (avoids sending event when users are selecting the command but not going to the end of it)

lhein replied:
yep, good point

apupier commented on src/extension.ts:
```diff
@@ -50,12 +52,44 @@ function sendOpenEvent() {
+		if (!file.toLowerCase().endsWith('.adm')) {
+			file += '.adm';
+		}
+		try {
+			const stat: vscode.FileStat = await vscode.workspace.fs.stat(vscode.Uri.file(file));
+			if (stat && stat.type === vscode.FileType.File) {
+				vscode.window.showErrorMessage(`The file ${file} already exists!`);
+				Promise.reject(new Error(`The file ${file} already exists!`));
+				return;
+			}	
+		} catch (error) {
+			// file doesn't exist - that's good
+		}		
```
can use a simpler method fs.existsSync  https://github.com/camel-tooling/vscode-camelk/blob/ed9c4e66feebb48d7a0c4f06087e321752288850/src/commands/NewIntegrationFileCommand.ts#L101

lhein replied:
I did not want to use that library but rather use the vscode internal fs api which does not have this function.

apupier replied:
it introduces more code but well why not. needs to handle these 2 comments then: https://github.com/jboss-fuse/vscode-atlasmap/pull/729/files#r756130524 and https://github.com/jboss-fuse/vscode-atlasmap/pull/729/files#r756131135

apupier replied:
in the VS Code samples, to check the existence they are using https://github.com/microsoft/vscode-extension-samples/blob/0b3a31bf2bdd388ac4fdc0ccea2fb1315abfe3e3/fsconsumer-sample/src/extension.ts#L27-L31

so means not checking for error type. So I guess that implicitly they defined that there is a single error possible

apupier commented on src/extension.ts:
```diff
@@ -50,12 +52,44 @@ function sendOpenEvent() {
+	sendCreateEvent();
+	const selectedWorkspaceFolder: vscode.WorkspaceFolder | undefined = await vscode.window.showWorkspaceFolderPick( {placeHolder: 'Please select the workspace folder in which the new file will be created.'} );
+	if (selectedWorkspaceFolder) {
+		const fileName: string = await vscode.window.showInputBox( {placeHolder: "Enter the name of the new AtlasMap file"});
+		var file: string = `${selectedWorkspaceFolder.uri.fsPath}/${fileName}`;
+		if (!file.toLowerCase().endsWith('.adm')) {
+			file += '.adm';
+		}
+		try {
+			const stat: vscode.FileStat = await vscode.workspace.fs.stat(vscode.Uri.file(file));
+			if (stat && stat.type === vscode.FileType.File) {
+				vscode.window.showErrorMessage(`The file ${file} already exists!`);
+				Promise.reject(new Error(`The file ${file} already exists!`));
+				return;
+			}	
+		} catch (error) {
+			// file doesn't exist - that's good
+		}		
+		await vscode.workspace.fs.writeFile(vscode.Uri.file(file), Buffer.from(''));
```
using a validator on the filed allows to have a faster feedback to users. For instance https://github.com/camel-tooling/vscode-camelk/blob/ed9c4e66feebb48d7a0c4f06087e321752288850/src/commands/NewIntegrationFileCommand.ts#L62

lhein replied:
that is a good improvement

apupier commented on src/extension.ts:
```diff
@@ -50,12 +52,44 @@ function sendOpenEvent() {
+	sendCreateEvent();
+	const selectedWorkspaceFolder: vscode.WorkspaceFolder | undefined = await vscode.window.showWorkspaceFolderPick( {placeHolder: 'Please select the workspace folder in which the new file will be created.'} );
+	if (selectedWorkspaceFolder) {
+		const fileName: string = await vscode.window.showInputBox( {placeHolder: "Enter the name of the new AtlasMap file"});
```
could be ncie to validate also that it doesn't contain invalid character for filename, see for example https://github.com/camel-tooling/vscode-camelk/blob/ed9c4e66feebb48d7a0c4f06087e321752288850/src/commands/NewIntegrationFileCommand.ts#L104

lhein replied:
good idea!

apupier commented on src/test/commands/atlasmap.file.create.test.ts:
```diff
@@ -0,0 +1,97 @@
+	let workspaceSelectorStub: sinon.SinonStub;
+	let fileNameInputStub: sinon.SinonStub;
+	
+	let testADMFileName: string = 'test.adm';
```
```suggestion
	const testADMFileName: string = 'test.adm';
```

apupier commented on src/test/commands/atlasmap.file.create.test.ts:
```diff
@@ -0,0 +1,97 @@
+	
+	let testADMFileName: string = 'test.adm';
+	let testADMFile: vscode.Uri;
+	let dummyADMFileName: string = 'dummy.adm';
```
```suggestion
	const dummyADMFileName: string = 'dummy.adm';
```

apupier reviewed (changes requested):
I think the functionality worth a note in the readme

apupier commented on src/extension.ts:
```diff
@@ -50,12 +52,44 @@ function sendOpenEvent() {
+				return;
+			}	
+		} catch (error) {
+			// file doesn't exist - that's good
```
the error type needs to be checked. Otherwise, it can be a completely another error thrown.

apupier replied:
this is not done in official example...
https://github.com/microsoft/vscode-extension-samples/blob/0b3a31bf2bdd388ac4fdc0ccea2fb1315abfe3e3/fsconsumer-sample/src/extension.ts#L30

apupier commented on src/extension.ts:
```diff
@@ -50,12 +52,44 @@ function sendOpenEvent() {
+		}
+		try {
+			const stat: vscode.FileStat = await vscode.workspace.fs.stat(vscode.Uri.file(file));
+			if (stat && stat.type === vscode.FileType.File) {
```
why checking for type of file? If it is a folder which exists with the name, we also need to detect it as an error

apupier commented on src/extension.ts:
```diff
@@ -50,12 +52,44 @@ function sendOpenEvent() {
+				Promise.reject(new Error(`The file ${file} already exists!`));
+				return;
+			}	
+		} catch (error) {
```
`error` variable can be removed as it is not used after

apupier commented on package-lock.json:
```diff
@@ -1,9419 +1,8 @@
 	"name": "atlasmap-viewer",
 	"version": "0.1.0",
-	"lockfileVersion": 2,
+	"lockfileVersion": 1,
```
seems you are using an older version of npm locally.
You will need to revert the package-lock.json and the change of dependency in package;json.
Upgrade npm
call again the npm install valid-filename@3.1.0

lhein replied:
reverted to v2

apupier commented on src/extension.ts:
```diff
@@ -50,12 +53,61 @@ function sendOpenEvent() {
+			}			
+		}
+	} finally {
+		sendCreateEvent();
```
note: Having it in the finally means the event is sent even when a user finally escape from the command and doesn't go to the end.

lhein replied:
yep, but he invoked the command. Don't we want to track that? Do we really want to track only successful invocations?

apupier replied:
in an ideal word, tracking each cases would be nice, with the step where the user abandoned or if there was an exception that prevented the command to go to the end.

As a first iteration, both (track all or only successful) are fine. Both are their advantages an drawbacks. In other Camel Tooling extensions, we went for the other case, to send for only successful ones.
(and another iteration might not be worthy until we have thousands of users)

lhein replied:
ok, only tracking the successful calls now to align with the other extensions.

lhein commented:
![createAndOpenADM](https://user-images.githubusercontent.com/435853/143276408-1a2f1818-b9d8-411f-a2f4-a65be0412d8d.gif)

apupier reviewed (changes requested):
remains the package-lock.json, see https://github.com/jboss-fuse/vscode-atlasmap/pull/736 to see the benefit on the diff with the steps I mentioned to update the dependencies
- revert package-lock.json to state before adding it
- remove valid-filename from package.json
- update local npm to version 7+ (for instance using nvm with command
nvm use lts/gallium)
- call nvm install valid-filename@3.1.0

The best will be to do it on this branch and to squash the commit that modifies the package.json
